src/goto-checker/solver_factory.cpp

```diff
@@ -178,12 +178,16 @@ static std::unique_ptr<SatcheckT>
 make_satcheck_prop(message_handlert &message_handler, const optionst &options)
 {
   auto satcheck = util_make_unique<SatcheckT>(message_handler);
-  if(options.is_set("write-solver-stats-to"))
+  if(auto hardness_collector = dynamic_cast<hardness_collectort *>(&*satcheck))
   {
-    satcheck->enable_hardness_collection();
-    satcheck->with_solver_hardness([&options](solver_hardnesst &hardness) {
-      hardness.set_outfile(options.get_option("write-solver-stats-to"));
-    });
+    if(options.is_set("write-solver-stats-to"))
```

Comment:
I think we should provide an error message to the user in case `--write-solver-stats-to` is used with a SAT solver that doesn't support hardness collection.